vkubelet/podcontroller.go

```diff
@@ -0,0 +1,339 @@
+
+	// Create an instance of PodController having a work queue that uses the rate limiter created above.
+	pc := &PodController{
+		server:       server,
+		podsInformer: server.podInformer,
+		podsLister:   server.podInformer.Lister(),
+		workqueue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "pods"),
+		recorder:     recorder,
+	}
+
+	// Set up event handlers for when Pod resources change.
+	pc.podsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(pod interface{}) {
+			if key, err := cache.MetaNamespaceKeyFunc(pod); err != nil {
+				runtime.HandleError(err)
+			} else {
+				pc.workqueue.AddRateLimited(key)
+			}
+		},
+		UpdateFunc: func(_, pod interface{}) {
+			if key, err := cache.MetaNamespaceKeyFunc(pod); err != nil {
+				runtime.HandleError(err)
+			} else {
+				pc.workqueue.AddRateLimited(key)
+			}
+		},
+		DeleteFunc: func(pod interface{}) {
+			if key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(pod); err != nil {
+				runtime.HandleError(err)
+			} else {
+				pc.workqueue.AddRateLimited(key)
+			}
+		},
+	})
+
+	// Return the instance of PodController back to the caller.
+	return pc
+}
+
+// Run will set up the event handlers for types we are interested in, as well as syncing informer caches and starting workers.
+// It will block until stopCh is closed, at which point it will shutdown the work queue and wait for workers to finish processing their current work items.
+func (pc *PodController) Run(ctx context.Context, threadiness int) error {
+	defer runtime.HandleCrash()
+	defer pc.workqueue.ShutDown()
+
+	// Wait for the caches to be synced before starting workers.
+	if ok := cache.WaitForCacheSync(ctx.Done(), pc.podsInformer.Informer().HasSynced); !ok {
+		return vkerrors.DeadlineWithMessage("failed to wait for caches to sync")
+	}
+
+	// Perform a reconciliation step that deletes any dangling pods from the provider.
+	// This happens only when the virtual-kubelet is starting, and operates on a "best-effort" basis.
+	// If by any reason the provider fails to delete a dangling pod, it will stay in the provider and deletion won't be retried.
+	pc.deleteDanglingPods(ctx)
+
+	// Launch two workers to process Pod resources.
+	log.G(ctx).Info("starting workers")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(pc.runWorker, time.Second, ctx.Done())
+	}
+
+	log.G(ctx).Info("started workers")
+	<-ctx.Done()
+	log.G(ctx).Info("shutting down workers")
+
+	return nil
+}
+
+// runWorker is a long-running function that will continually call the processNextWorkItem function in order to read and process an item on the work queue.
+func (pc *PodController) runWorker() {
+	for pc.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem will read a single work item off the work queue and attempt to process it,by calling the syncHandler.
+func (pc *PodController) processNextWorkItem() bool {
+	obj, shutdown := pc.workqueue.Get()
```

Comment:
Some tracing in this function (and recursively) would be nice.

Reply:
I have very little experience with tracing. Adding #404, I thought maybe we could postpone this to a later phase (pre-1.0).

Also, I am trying to keep the size of this PR as little as possible so it's easier to review.

Does this convince you? If not, I'll try and see what I can do as part of this PR.

Reply:
I can do this as part of my next task, work on #348. WDYT?

Reply:
It's pretty easy to add. I'd like to get it in here if we can.

You create a span (`trace.StartSpan(ctx, <a span name>)`), the span should end at the end of the function (typically, and should be fine here by calling `defer span.End()`)
The span context is passed through a `context.Context`.
There should already be a helper to add pod attributes to the span.

If the function ends in an error you can set a span status `span.SetStatus(ocstatus.FromError(err))`.

Reply:
I'll look into it tomorrow morning my time.

Reply:
Done in 1a4b5b0 and 5ea3b72.

![vk_create_pod_trace](https://user-images.githubusercontent.com/1752631/49298168-9bcd1700-f4b4-11e8-92f8-aab3a7da7bb6.png)
![vk_delete_pod_trace](https://user-images.githubusercontent.com/1752631/49298169-9bcd1700-f4b4-11e8-9836-0380441d4266.png)